presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBatch.java

```diff
@@ -179,6 +243,70 @@ public PrestoSparkRowBatch build()
                     rowOffsets,
                     totalSize);
         }
+
+        private PrestoSparkRowBatch createGroupedRowBatch()
+        {
+            RowIndex rowIndex = RowIndex.create(rowCount, partitionCount, rowPartitions);
+            byte[] data = sliceOutput.getUnderlyingSlice().byteArray();
+
+            DynamicSliceOutput output = new DynamicSliceOutput((int) (totalSize * 1.2f));
+            int expectedEntriesCount = (int) ((totalSize / targetAverageRowSizeInBytes) * 1.2f);
+            int[] entryOffsets = new int[expectedEntriesCount];
+            int[] entryPartitions = new int[expectedEntriesCount];
+            int entriesCount = 0;
+            for (int partition = REPLICATED_ROW_PARTITION_ID; partition < partitionCount; partition++) {
+                while (rowIndex.hasNextRow(partition)) {
```

Comment:
is this `while` try to skip empty partition? -- in that case can it be an `if` statement instead ? :)